### PR TITLE
[Gen 4] Headbutt (#71)

### PR DESCRIPTION
**Pull request by Real96**
Added HGSS Headbutt in Wild Generator and Searcher (#71)

### CODE REVIEW

File: Source/Core/Enum/Encounter.hpp [hunk 1]
```diff
@@ -35,8 +35,9 @@ enum class Encounter : u8
     SuperRod,
     SpecialSuperRod,
     Static,
+    SafariZone,
```
Comment (Admiral-Fish) [inline Source/Core/Enum/Encounter.hpp:38]: Why was safarizone added back? This is just deadcode right now

File: Source/Core/Gen4/Encounters4.cpp [hunk 2]
```diff
@@ -31,6 +30,27 @@
 
 namespace Encounters4
 {
+    bool getHeadbuttSpecialFlag(Game game, int location)
```
Comment (Admiral-Fish) [inline Source/Core/Gen4/Encounters4.cpp:33]: Please move this function down below getEncounters()

File: Source/Core/Gen4/Encounters4.cpp [hunk 3]
```diff
@@ -224,6 +244,32 @@ namespace Encounters4
                 }
             }
 
+            if (encounter == Encounter::Headbutt)
+            {
+                int offset = 0;
+
+                for (int i = 0; i < 59; i++)
```
Comment (Admiral-Fish) [inline Source/Core/Gen4/Encounters4.cpp:251]: Not a fan of this approach. It doesn't match how the rest of the tables are handled.

Would rather have
```
for (size_t offset = 0; offset < size; ) {
// Handle table

offset += (num based on special tree);
}
```

File: Source/Core/Gen4/Encounters4.hpp [hunk 4]
```diff
@@ -20,6 +20,7 @@
 #ifndef ENCOUNTERS4_HPP
 #define ENCOUNTERS4_HPP
 
+#include <Core/Enum/Game.hpp>
```
Comment (Admiral-Fish) [inline Source/Core/Gen4/Encounters4.hpp:23]: Don't include this header unless absolutely necessary at the header level.

Just have a `enum class Game : u32;`

File: Source/Core/Util/EncounterSlot.cpp [hunk 5]
```diff
@@ -100,7 +100,9 @@ namespace EncounterSlot
             return calcSlot<5>(compare, std::array<u8, 5> { 60, 90, 95, 99, 100 });
         case Encounter::BugCatchingContest:
             return calcSlot<10, true>(compare, std::array<u8, 10> { 80, 60, 50, 40, 30, 20, 15, 10, 5, 0 });
-        case Encounter::HeadButt:
+        case Encounter::SafariZone:
+            return compare % 10;
```
Comment (Admiral-Fish) [inline Source/Core/Util/EncounterSlot.cpp:104]: Why was this added back. It's dead code right now

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]
General question: does changing `time` to `modifier` have any downside?

File: Source/Core/Enum/Encounter.hpp (hunk 1 above)
Comment (Real96) [reply Source/Core/Enum/Encounter.hpp]: Removed

File: Source/Core/Gen4/Encounters4.cpp (hunk 2 above)
Comment (Real96) [reply Source/Core/Gen4/Encounters4.cpp]: > Please move this function down below getEncounters()

done

File: Source/Core/Gen4/Encounters4.cpp (hunk 3 above)
Comment (Real96) [reply Source/Core/Gen4/Encounters4.cpp]: > Not a fan of this approach. It doesn't match how the rest of the tables are handled.
> 
> Would rather have
> 
> ```
> for (size_t offset = 0; offset < size; ) {
> // Handle table
> 
> offset += (num based on special tree);
> }
> ```

done

File: Source/Core/Gen4/Encounters4.hpp (hunk 4 above)
Comment (Real96) [reply Source/Core/Gen4/Encounters4.hpp]: > Don't include this header unless absolutely necessary at the header level.
> 
> Just have a `enum class Game : u32;`

done

File: Source/Core/Util/EncounterSlot.cpp (hunk 5 above)
Comment (Real96) [reply Source/Core/Util/EncounterSlot.cpp]: Removed

Comment (SteveCookTU) [reply Source/Core/Util/EncounterSlot.cpp]: This was my bad, somehow was leftover from merging in updated items

File: Source/Forms/Gen4/Wild4.cpp [hunk 6]
```diff
@@ -254,6 +261,42 @@ void Wild4::updatePokemonSearcher()
     }
 }
 
+void Wild4::updateTreesTypeGenerator()
+{
+    if (Encounters4::getHeadbuttSpecialFlag(currentProfile->getVersion(),
+                                            encounterGenerator[ui->comboBoxGeneratorLocation->currentData().toInt()].getLocation()))
+    {
+        if (ui->comboBoxGeneratorTreesType->findData(toInt(HeadbuttType::Special)) < 0)
```
Comment (Admiral-Fish) [inline Source/Forms/Gen4/Wild4.cpp:269]: `< 0` -> `== -1`

According to the docs: `Returns the index of the item containing the given data for the given role; otherwise returns -1.`

File: Source/Forms/Gen4/Wild4.cpp [hunk 7]
```diff
@@ -254,6 +261,42 @@ void Wild4::updatePokemonSearcher()
...
+
+    ui->comboBoxGeneratorTreesType->setCurrentIndex(0);
+}
+
+void Wild4::updateTreesTypeSearcher()
+{
+    if (Encounters4::getHeadbuttSpecialFlag(currentProfile->getVersion(),
+                                            encounterSearcher[ui->comboBoxSearcherLocation->currentData().toInt()].getLocation()))
+    {
+        if (ui->comboBoxSearcherTreesType->findData(toInt(HeadbuttType::Special)) < 0)
```
Comment (Admiral-Fish) [inline Source/Forms/Gen4/Wild4.cpp:287]: `< 0` -> `== -1`

According to the docs: `Returns the index of the item containing the given data for the given role; otherwise returns -1.`

File: Source/Forms/Gen4/Wild4.cpp [hunk 8]
```diff
@@ -500,14 +566,23 @@ void Wild4::searcherEncounterIndexChanged(int index)
         case Encounter::RockSmash:
             t = { "0", "1" };
             break;
+        case Encounter::Headbutt:
+            t = { "0", "1", "2", "3", "4", "5" };
```
Comment (Admiral-Fish) [inline Source/Forms/Gen4/Wild4.cpp:570]: Likely missing the `flag = true;` here

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]

File: Source/Forms/Util/EncounterLookup.cpp [hunk 9]
```diff
@@ -121,8 +122,9 @@ std::set<std::pair<u16, QString>> EncounterLookup::getEncounters4(Game game, u16
     std::vector<Profile4> profiles;
 
     // Encounter variables to iterate through
-    auto types = { Encounter::Grass,   Encounter::RockSmash,         Encounter::OldRod, Encounter::GoodRod, Encounter::SuperRod,
-                   Encounter::Surfing, Encounter::BugCatchingContest };
+    auto types = { Encounter::Grass,   Encounter::RockSmash,          Encounter::OldRod,  Encounter::GoodRod, Encounter::SuperRod,
+                   Encounter::Surfing, Encounter::BugCatchingContest, Encounter::Headbutt };
+    auto treeTypes = { HeadbuttType::Normal1, HeadbuttType::Normal2, HeadbuttType::Special };
```
[the remark this answers is not in the document]
Comment (Real96) [reply Source/Forms/Util/EncounterLookup.cpp:127]: Yes, removed

File: Source/Forms/Gen4/Wild4.cpp (hunk 6 above)
Comment (Real96) [reply Source/Forms/Gen4/Wild4.cpp:269]: done

File: Source/Forms/Gen4/Wild4.cpp (hunk 7 above)
Comment (Real96) [reply Source/Forms/Gen4/Wild4.cpp:287]: done

File: Source/Forms/Gen4/Wild4.cpp (hunk 8 above)
Comment (Real96) [reply Source/Forms/Gen4/Wild4.cpp:570]: added

**Review by Admiral-Fish — APPROVED** [review]
